src/setfit/onnx.py

```diff
@@ -0,0 +1,281 @@
+    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)
+
+
+class OnnxSetFitModel(torch.nn.Module):
+    """A wrapper around SetFit model body, pooler, and model head which makes ONNX exporting easier.
+
+    This wrapper allows us to create a nn.Module with different levels of connectivity. We can set
+    model_body and pooler and have a Module which maps inputs to embeddings or we can set all three
+    and have a model which maps inputs to final predictions. This is useful because torch.onnx.export
+    will work with a nn.Module.
+
+    Attributes:
+        model_body (`PreTrainedModel`): The pretrained model body of a setfit model.
+        pooler (`nn.Module` or `Callable[[torch.Tensor], torch.Tensor]`): The callable function that can
+            map  tensors of shape (batch, sequence, embedding_dim) to shape (batch, embedding_dim).
+        model_head: (`torch.nn.Module` or `LogisticRegression`): The model head from the pretrained SetFit model.
+            if it is None then the resulting OnnxSetFitModel forward will return embeddings instead of predictions.
+    """
+
+    def __init__(
+        self,
+        model_body: PreTrainedModel,
+        pooler: Optional[Union[nn.Module, Callable[[torch.Tensor], torch.Tensor]]] = None,
+        model_head: Optional[Union[torch.nn.Module, LogisticRegression]] = None,
+    ):
+        super().__init__()
+
+        self.model_body = model_body
+        if pooler is None:
+            self.pooler = mean_pooling
+        else:
+            self.pooler = pooler
+        self.model_head = model_head
+
+    def forward(self, input_ids: torch.Tensor, attention_mask: torch.Tensor, token_type_ids: torch.Tensor):
+        hidden_states = self.model_body(input_ids, attention_mask, token_type_ids)
+        hidden_states = {"token_embeddings": hidden_states[0], "attention_mask": attention_mask}
+
+        embeddings = self.pooler(hidden_states)
+
+        # If the model_head is none we are using a sklearn head and only output
+        # the embeddings from the setfit model
+        if self.model_head is None:
+            return embeddings
+
+        # If head is set then we have a fully torch based model and make the final predictions
+        # with the head.
+        out = self.model_head(embeddings)
+        return out
+
+
+def export_onnx_setfit_model(setfit_model: OnnxSetFitModel, inputs, output_path, opset):
+    """Export the OnnxSetFitModel.
+
+    This exports the model created by the OnnxSetFitModel wrapper using torch.onnx.export.
+
+    Args:
+        setfit_model (`OnnxSetFitModel`): The OnnxSetFitModel we want to export to .onnx format.
+        inputs (`Dict[str, torch.Tensor]`): The inputs we would hypothetically pass to the model. These are
+            generated using a tokenizer.
+        output_path (`str`): The local path to save the onnx model to.
+        opset (`int`): The ONNX opset to use for the export.
```

Comment:
What do you think about setting a good default value for `opset` since it might be annoying for users to have to look it up

Reply:
That's a great suggestion! I've had 12 work well so I'll set that to start : )